### ISSUE TITLE
Edge not displayed

### ISSUE THREAD

**Issue by ludenge** [xyflow/xyflow#4888]
### What platform were you using when you found the bug?

- React Flow / Svelte Flow version: 12.3.6
- Browser and version:  130.0.6723.119（正式版本） (arm64)
- OS and version: Mac 14.1.2 (23B92)

### Live code example

https://codesandbox.io/p/devbox/suspicious-jackson-3kcxkx?file=%2FButtonEdge.tsx&workspaceId=ws_PPWfBAbPnWvPVQAgAgALv9

### Describe the Bug

edge not displayed

### Steps to reproduce the bug or issue

You need to keep refreshing the codesandbox, and it might reappear.
(If you still can't reproduce it, you can try to trigger the hot update and refresh it again.)

### Expected behavior

Show edge correctly

### Screenshots or Videos

![20241215165912_rec_](https://github.com/user-attachments/assets/b257e02d-4117-41c2-bacf-b58a4d1c3cde)

**Comment by ludenge** [comment]
In fact, this problem is definitely reproducible in my own project, but I've spent a lot of time and still can't isolate the code that consistently reproduces it.

**Comment by bcakmakoglu** [comment]
Your sandbox isn't available.
<img width="632" alt="image" src="https://github.com/user-attachments/assets/e1a6a788-5a73-42d5-9927-69781eeaaa35" />

**Comment by ludenge** [comment]
> Your sandbox isn't available. <img alt="image" width="632" src="https://private-user-images.githubusercontent.com/78412429/395845749-e1a6a788-5a73-42d5-9927-69781eeaaa35.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyNTU3MDMsIm5iZiI6MTczNDI1NTQwMywicGF0aCI6Ii83ODQxMjQyOS8zOTU4NDU3NDktZTFhNmE3ODgtNWE3My00MmQ1LTk5MjctNjk3ODFlZWFhYTM1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMTUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjE1VDA5MzY0M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ5N2IwNzU2OTBiY2FkMTNjMTA4ZTg2ZWFmNjBmMTQ5YzdhNDAyMWZlMDA2OWU1YWNkYzEyYzc3MjNkMDVhYTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.GMIjusc6s822m532jnUkHya3Lw1waO7bz8Y6MCI39_Q">

Sorry, please take a look again. Or visit the following link https://codesandbox.io/p/live/0618ee89-7f33-4fc7-9dec-4f3b779a98f1

**Comment by bcakmakoglu** [comment]
Seems like the useEffect you call in your `AddNode` that's supposed to do the initial data update creates some issue that causes edges not to be rendered, I'm not sure *what* exactly the problem is 🤷‍♂️

```ts
  // removing this effect fixes the issue
  useEffect(() => {
    onValueChange(data);
  }, []);
```

I'm not sure why you need this useEffect anyway? You get some data in props, which is your initial data, only to then update your node again with the same data? What's the reasoning for that?

**Comment by ludenge** [comment]
I wrote this code to reproduce the problem. In actual applications, we will update some meaningful values, and the logic cannot be deleted.

**Comment by ludenge** [comment]
I also encountered a problem, when I clicked edge to add a node, the edge disappeared, but I couldn't reproduce this issue in Codesandbox. It's probably caused by the same reason. I hope you can check it out.

**Comment by moklick** [comment]
Seems to be a race condition with updateNodeData called initially 🤔 We will investigate (does this also happen with React 18?)

**Comment by Jonasb8** [comment]
Hi 👋

I'm experiencing an issue with dynamically adding nodes and edges to my flow.

When I trigger an action that should add a new node along with an associated edge, it sometimes works as expected. However, in some cases, all existing edges disappear from the view—even though there are no errors in the console.

Additionally, if I click the trigger again, some of the nodes start disappearing as well. Despite this, when I log the edge data to the console, all the expected edges are still present in the data structure—they're just not rendering.

I've attached a short video clip demonstrating the issue. Let me know if code snippets would help—I'd be happy to provide them.

https://github.com/user-attachments/assets/f0ef328e-31c1-4c4a-827c-e584b171107c

**Comment by moklick** [comment]
This looks odd @Jonasb8 . How are you setting nodes and edges?

**Comment by Jonasb8** [comment]
here is a simplified version of my code : when clicking the node with the plus icon I trigger an API call
when this call is sucessfull I want to display a new node

```
export function useProjectFlowNodes() {
  const { project } = useProjectStore((state) => ({ project: state.project }));
  const {setProject } = useProjectStore();

  const [loadingNodeId, setLoadingNodeId] = useState<string | null>(null);
  const [nodes, setNodes, onNodesChange] = useNodesState<Node>([]);
  const [edges, setEdges, onEdgesChange] = useEdgesState<Edge>([]);

  const { mutate: callApi } = useCallApi({
    onSuccess: (data) => {
      if (data) {
        setProject(data);
      }
    },
    onError: () => {
      console.error('Error sending message');
    },
  });

  const handleClickOnNode = useCallback((nodeId: string) => {
    setLoadingNodeId(nodeId);
    callApi({ ... });
  }, [callApi);

  useEffect(() => {
    if (projectHistory) {

      const projectScope = project?.state

      const allNodes: Node[] = [];
      const allEdges: Edge[] = [];

      projectScope.forEach((item, index) => {
        item.causes?.forEach((nestedIndex) => {
          const node = {
            id: `node-${index}-${nestedIndex}`,
            type: 'customNode1',
            position: { ... },
            data: {
              onClick: () => handleClickOnNode(`node-${index}-${nestedIndex}`),
            },
          };

          allNodes.push(node);
          allEdges.push({
            id: `edge-nested-to-node-${index}-${nestedIndex}`,
            source: `nested-causal-node-${index}-${nestedIndex}`,
            target: node.id,
            type: 'smoothstep',
          });
        });
      });

      setNodes(allNodes);
      setEdges(allEdges);
    }
  }, [projectHistory, project, handleClickOnNode, loadingNodeId]);

  return { nodes, edges, onNodesChange, onEdgesChange, callApi };
}

**Comment by bcakmakoglu** [comment]
```
source: `nested-causal-node-${index}-${nestedIndex}`
```

Where does this id come from? Your code snippet doesn't seem to have any reference to a node that would have this id.
I see these ids on nodes

```
id: `node-${index}-${nestedIndex}`,
```

```
id: `new-node-${index}-${nestedIndex}`,
```

Are you sure that "nested-causal-node-..." exists?

**Comment by Jonasb8** [comment]
yes it does, I didn't provide it in the previous snippet because the whole file is too big
here is my full code if it helps : https://gist.github.com/Jonasb8/cc09f1f641bb4356b65259d76e0146d5

**Comment by bcakmakoglu** [comment]
Mh... yeah a bit complicated to go through this.

Maybe you can confirm that these potential problems are *not* present in your node configuration:

1. Are all custom node types properly mapped in the `nodeTypes` object? Have you ensured you didn't misspell anything?
2. Did you check and ensure that no ids are *duplicated*?
3. Did you check and ensure that node positions are valid (not NaN or anything like that)?
4. Are the `sourceHandle` and `targetHandle` ids properly assigned inside of your custom node types? Ensure that these ids actually are passed to `<Handle>` components (for example you use `sourceHandle: 'project-analysis-handle'` - is this id passed as `<Handle id="project-analysis-handle" ...>`?

Other than that, are you sure that you want to run this whole effect when `clickedSubCause` changes? 
Maybe it'd be better to extract the logic for that out of this whole loop?

**Comment by Jonasb8** [comment]
Nope, none of those problems are there, I am able to render all nodes and edges 80% of the time 
but maybe since I am re-rendering a large amount of nodes I can happen that some of the nodes are not yet rendered when the edges are being rendered and this cause the edge's display to fail 🤔

**Comment by lucasgelfond** [comment]
Bumping this to say I'm having similar issues. Previously, when stepping through various parts of layout rendering, the edges would also show non-deterministically; does seem like a race condition, and I only get bugs maybe every ~fifth pass. 

The code @ludenge posted is a way more minimal / isolated reproduction (thank you!!). I followed all of @bcakmakoglu's advice above on the minimal reproduction, and all of them aren't present:

1. single type, spelled correctly
2. visually inspect node and edge ids, all good
3. node positions are passed explicitly, all are legit 
4. console.log-ed out the handle ids, and they are in factg correct / passed correctly

any idea what to do / fix?

**Comment by moklick** [comment]
We need a reproduction for this. @lucasgelfond are you using `updateNodeData` or any other helpers of `useReactFlow` regarding nodes or edges?

**Comment by lucasgelfond** [comment]
The Codesandbox linked above reproduces the issue, @ludenge 's: https://codesandbox.io/p/devbox/suspicious-jackson-3kcxkx?file=%2FButtonEdge.tsx&workspaceId=ws_PPWfBAbPnWvPVQAgAgALv9 . Some percentage of the time (maybe 10-15%) the edges don't show up on render.

I can get you a minimal reproduction of mine, will take a bit but can do now! I'm using a bunch of `useReactFlow` stuff: `setCenter`, `setNodes`, and `setEdges` primarily.

**Comment by LoicTryToDev** [comment]
Good evening everyone.

I had the same problem. I solved it by fixing my handle configuration in my customNodes.

I only had "handle" of "source" type
```
 <Handle
 id="handle-bottom"
 type="target"
 position={Position.Top}
 />
```
By correctly adding a "target" for my target type nodes, the edges appeared correctly, without duplicates or "bugs".
if you have a large number of elements, couldn't there be one with the wrong type (customType), a "type" that doesn't have the proper "handle".
I no longer have access to your repo, so I wasn't able to look at the code.

In your code above: 
```
 allEdges.push({
 id: `edge-nested-to-node-${index}-${nestedIndex}`,
 source: `nested-causal-node-${index}-${nestedIndex}`,
 target: node. id,
 type: 'smoothstep',
 });
```
I'm assuming that the node with the id "nested-causal-node-${index}-${nestedIndex}" exists and that it has the type "customNode2"
the target, so node (node.id) has the type "customNode1".

**You need to check that ‘customNode2’ contains a ‘source’ type handle and that your “customNode1” contains a ‘target’ type handle.**


Good evening to you all.

**Comment by RobertMic** [comment]
👋 We've been seeing this in our app and I did some debugging. Going to write down my findings here in case it's helpful. I tried my best to try to understand how `ReactFlow`'s internals work to debug this but ultimately I am not that familiar with this code, so hopefully this is correct and I'm not misinterpreting what I saw.

I did try to make a minimal example that replicates this but failed. So clearly there is something I am not understanding here. It does seem though, that the code can get into a state where `measured` is set on an internal node but `internals.handleBounds` isn't, and whatever mechanism is meant to calculate that fails to do so.

## TLDR;

I believe this is caused by a race between the `ResizeObserver` and the `StoreUpdater` and it is made more evident when the React app is slow to render. Since the `ResizeObserver` writes to the `zustand` store outside of the React lifecycle, sometimes a React render can use "old" React data but "new" `zustand` data when `StoreUpdater` ultimately calls `adoptUserNodes`.

## Assumptions

- `ResizeObserver` runs in the main thread, like the app's React code
- The browser can queue React code because it needs to schedule the `ResizeObserver`
- The user is using a controlled flow where they have their own representation of a `node` which is propagated to the `ReactFlow` component via the `nodes` prop
- The user's `nodes` representation is a simple `useState`. The user calls `setUserRepresentationNodes` when they make an update on their end (server mutation happens, etc)
- The user listens to `onNodesChange` and uses `applyNodeChanges` to apply `dimensions` changes to their node representation

## Sequence

- Initial render happens, the `internalNode` lookup gets populated with nodes that have no `handleBounds` and no `measured.height/width`. The DOM either isn't fully painted or the `ResizeObserver` hasn't run yet.
- The user's slow `useEffect` runs and updates the `nodes` property on the `ReactFlow` component. If the timing is right, the browser will schedule the `ResizeObserver` callback to run after this code is done but BEFORE the `StoreUpdater` sees the changes in its `useEffect`
- `ResizeObserver` runs and calls `updateNodeInternals`. This calculates the `handleBounds` and `measured` properties as expected on the internal nodes. A call to `onNodesChange` happens which causes the app to update `nodes` again, but with a `measured` property. (This gets queued AFTER the previous `nodes` update)
- `StoreUpdater`'s `useEffect` runs and calls `adoptUserNodes` for the **first** `nodes` update  (the one where `measured` is still `undefined`). At this point the `zustand` store is correct, the `internalNode` has `measured` and `handleBounds`, but the `userNode` is old and seen to have no `measured`. The "re-initialize" logic gets hit which sets the `internalNode`'s `handleBounds` and `measured.height/width` to `undefined`
- `StoreUpdater`'s `useEffect` runs and calls `adoptUserNodes` for the **second** `nodes` update which **does** have `measured` set. It sets the `internalNode` to have a `measured` property but uses the now incorrect `handleBounds` on `internalNode`, which is `undefined`
- Nothing at this point seems to recognize that `measured` is set, but `internalNode.internals.handleBounds` **isn't**. This causes any edge to see the node as not initialized so it doesn't draw anything. New nodes added are correctly initialized but existing nodes don't seem to get re-initialized ever

This sequence can happen occasionally in our app. I made it artificially slow by putting an empty `for` loop with 50 million iterations in the "slow useEffect", this causes the bug to appear almost consistently.

## Logs

Here's a video and some screenshots of my debugging that I think proves the above is happening.

The logpoints I put in the code:

One set in `adoptUserNodes`. This one prints `true` when the "re-initialize" logic is true (`userNode.measured` is `undefined`, but `internalNode.internals.handleBounds` isn't) and prints out the `userNode.measured` and `internalNode.internals.handleBounds`

<img width="990" height="747" alt="Image" src="https://github.com/user-attachments/assets/5f2c4075-bb06-48a7-b680-b1b2d7daaf33" />

One set of logs in the `ResizeObserver` to show that it is running

<img width="765" height="453" alt="Image" src="https://github.com/user-attachments/assets/17cb201d-7bef-4f1b-be5a-a5a96d4bb6e4" />

One set of logs in our app code (which I won't show here) that logs when the "slow `useEffect`" runs. And another set in our app code that logs when we see a `dimensions` change happen.

I then keep refreshing the page until I see the bug:

<img width="1356" height="973" alt="Image" src="https://github.com/user-attachments/assets/e65b8dc4-92b7-4ae3-b301-44e103bb348c" />

- The first two `adoptUserNodes` logs are the initialization of the app
- The `Slow set` log is our slow `useEffect` running with an update. It writes two nodes with no `measured`
- The `ResizeObserver` log then runs, sets the `measured` and `internalBounds`. It also calls `onNodesChange` which causes us to write two new nodes with a `measured`. The `Dimensions change` log is on the app end, it's the `onNodesChange` callback running
- The two `adoptUserNodes` logs then show that its seeing the first `nodes` update with no `measured` set, but it _is_ seeing the `zustand` store update where the internal node has a `measured` and `handleBounds`
- Something on our end causes another call to `Slow set`, but it clearly has `measured` at this point
- `adoptUserNodes` runs again with the update where `userNode.measured` is set, but at this point `internalNode.internals.handleBounds` is `undefined` so it sets that again
- A bunch more re-renders happen but each time nothing seems to update the `internalNode.internals.handleBounds` so it always stays `undefined`
- The app finally settles and no more rendering is being done, the edges remain hidden

And finally here's a video of it happening

https://github.com/user-attachments/assets/c8dbc062-15a7-43f9-8794-9cef6cc7d9f8

**Comment by RobertMic** [comment]
I also ended up making this component and rendering it alongside the `ReactFlow` component (but under the provider). It seems to do an ok job at detecting when the edges don't render on our app and triggering an update on the `ReactFlow` side

```tsx
import {useStore, useUpdateNodeInternals} from '@xyflow/react';
import type {InternalNodeBase} from '@xyflow/system';
import {useEffect} from 'react';

type Slice = InternalNodeBase[];

const buggedNodesSelector: Parameters<typeof useStore<Slice>>[0] = ({
  nodeLookup,
}) => {
  const nodes = Array.from(nodeLookup.values());
  return nodes.filter(
    node =>
      node.measured?.height != null &&
      node.measured?.width != null &&
      node.internals.handleBounds === undefined,
  );
};

const equalityFn: Parameters<typeof useStore<Slice>>[1] = (
  prevSlice,
  nextSlice,
) => {
  return (
    prevSlice.length === nextSlice.length &&
    prevSlice.every((node, index) => node.id === nextSlice[index].id)
  );
};

export function EdgeBugHack() {
  const buggedNodes = useStore(buggedNodesSelector, equalityFn);
  const updateNodeInternals = useUpdateNodeInternals();

  useEffect(() => {
    if (buggedNodes.length > 0) {
      for (const node of buggedNodes) {
        updateNodeInternals(node.id);
      }
    }
  }, [buggedNodes, updateNodeInternals]);

  return null;
}
```
